File: App/Resources/AlgorithmDetails/basenmaxheapsort/basenmaxheapsort.py
BASE = 4

def sift_down(arr, node, stop):
  left = node * BASE + 1
  if left >= stop:
    return
  max_index = left
  i = left + 1
  while i < left + BASE and i < stop:
    if arr[max_index] < arr[i]:
      max_index = i
    i += 1
  if arr[node] < arr[max_index]:
    arr[node], arr[max_index] = arr[max_index], arr[node]
    sift_down(arr, max_index, stop)

def sort(arr):
  n = len(arr)
  for i in range(n - 1, -1, -1):
    sift_down(arr, i, n)
  for end in range(n - 1, 0, -1):
    arr[0], arr[end] = arr[end], arr[0]
    sift_down(arr, 0, end)

if __name__ == "__main__":
  array = [0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56]
  sort(array)
  print(array)

File: App/Resources/AlgorithmDetails/basenmaxheapsort/basenmaxheapsort.go
package main

import (
	"fmt"
)

const base = 4

func siftDown(arr []int, node int, stop int) {
	left := node*base + 1
	if left >= stop {
		return
	}
	maxIndex := left
	for i := left + 1; i < left+base && i < stop; i++ {
		if arr[maxIndex] < arr[i] {
			maxIndex = i
		}
	}
	if arr[node] < arr[maxIndex] {
		arr[node], arr[maxIndex] = arr[maxIndex], arr[node]
		siftDown(arr, maxIndex, stop)
	}
}

func sort(arr []int) []int {
	n := len(arr)
	for i := n - 1; i >= 0; i-- {
		siftDown(arr, i, n)
	}
	for end := n - 1; end > 0; end-- {
		arr[0], arr[end] = arr[end], arr[0]
		siftDown(arr, 0, end)
	}
	return arr
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56}
	fmt.Println(sort(array))
}
